co2calculator/calculate.py

```diff
@@ -21,78 +24,149 @@
+
     """
-    # NOTE: Tests fail for 'cng'  as `fuel_type` (IndexError)
+    coords = []
+    for loc in DistanceRequest.stops:
+        loc_name, loc_country, loc_coords, _ = geocoding_structured(loc.dict())
+        coords.append(loc_coords)
+    distance = get_route(coords, "driving-car")
+    return distance
 
-    transport_mode = "car"
 
-    # Set default values
-    if passengers is None:
-        passengers = 1
-        warnings.warn(
-            f"Number of car passengers was not provided. Using default value: '{passengers}'"
-        )
-    if size is None:
-        size = "average"
-        warnings.warn(f"Size of car was not provided. Using default value: '{size}'")
-    if fuel_type is None:
-        fuel_type = "average"
-        warnings.warn(
-            f"Car fuel type was not provided. Using default value: '{fuel_type}'"
-        )
-    # Check if distance of stops provided
-    if distance is None and stops is None:
-        raise ValueError(
-            "Travel parameters missing. Please provide either the distance in km or a list of"
-            "dictionaries of travelled locations."
+def trip_detour(request: DistanceRequest) -> Kilometer:
+    """
+    Function to calculate the distance of a trip using the distance as the crow flies and detour coefficients/constants
+
+    :param request: validated input data
+    :return: Distance in kilometer
+    :rtype: Kilometer
+    """
+    distance = 0
+    coords = []
+    for loc in DistanceRequest.stops:
+        loc_name, loc_country, loc_coords, _ = geocoding_structured(loc.dict())
+        coords.append(loc_coords)
+    for i in range(0, len(coords) - 1):
+        # compute great circle distance between locations
+        distance += haversine(
+            coords[i][1], coords[i][0], coords[i + 1][1], coords[i + 1][0]
         )
-    elif distance is None:
-        coords = []
-        for loc in stops:
-            loc_name, loc_country, loc_coords, _ = geocoding_structured(loc)
-            coords.append(loc_coords)
-        distance = get_route(coords, "driving-car")
+    distance = apply_detour(
+        distance, transportation_mode=DistanceRequest.transportation_mode
+    )
+
+    return distance
+
+
+class Car(BaseModel):
+    """Vehicle type Car"""
+
+    @enum.unique
+    class Size(str, enum.Enum):
+        """Available sizes for cars"""
+
+        SMALL = "small"
+        MEDIUM = "medium"
+        LARGE = "large"
+        AVERAGE = "average"
+
+    @enum.unique
+    class FuelType(str, enum.Enum):
+        """Available fuel types for cars"""
+
+        DIESEL = "diesel"
+        GASOLINE = "gasoline"
+        CNG = "cng"
+        ELECTRIC = "electric"
+        HYBRID = "hybrid"
+        PLUG_IN_HYBRID = "plug-in_hybrid"
+        AVERAGE = "average"
+        HYDROGEN = "hydrogen"
+
+    size: Size
+    fuel_type: FuelType
+
+
+class CarEmissionRequest(EmissionRequest):
+    """Request-contract to calculate co2 emissions"""
+
+    passengers: int
```

Comment:
Just FYI, not needed: `pydantic` comes with a lot of useful own types. E.g., `PositiveInt`. The model will validate that no one asked for -1 passengers.